### DISCUSSION TITLE
How MasterAPP transfers displacement information to SubAPP

### DISCUSSION THREAD

**Discussion by js-jixu** [idaholab/moose#22285] [Q&A General] [answered]
Hi, experts.

There is an initial mesh file, a MasterApp input file and a SubApp input file. First, MaterApp uses the initial mesh file to calculate, and calculates disp_x/y/z. Then I hope that when SubApp calculates it can use the displaced mesh calculated by MasterApp.

I am worried that my expression might not be very clear, so I would like to give a concrete example. MasterApp will calculate disp_x/y/z and SubApp will calculate temperature distribution. The displacement of the mesh affects the temperature distribution, so I want to calculate the temperature distribution on the latest displacement mesh. The temperature distribution also causes mesh displacement through thermal expansion:

![1](https://user-images.githubusercontent.com/99813361/193619254-dae260ea-ffd4-4a90-8366-54f3f037d3d3.png)

I thought of two ways. The first is to use the latest output file of the MasterApp as the mesh file of the SubApp. SubApp needs to use the latest file every time, can this be done?

The second method is to set auxiliary variables of disp_x/y/z in SubApp. MasterApp passes disp_x/y/z information to SubApp in Transfers. Then SubApp applies the disp_x/y/z information to the initial mesh file, and SubApp can also use the displaced mesh. The difficulty with this approach is how to apply the disp_x/y/z information to the initial mesh file. Does using use_displaced_mesh=true in SubApp do the trick?

**Comment by friedmud** [comment, marked as answer]
Your "second method" is the right one.  In MOOSE `disp_x/y/z` are field variables, just like any other variable.  Therefore they can be transferred just like any other variable.

In the sub-app, all you need to do is add `disp_x,y,z` as Auxiliary variables and set `displacements = 'disp_x disp_y disp_z'` in the mesh block.  You can see the documentation for the `displacements` option under `Mesh Displacements` here: https://mooseframework.inl.gov/syntax/Mesh/index.html

Note: if you are using one of the fancier Actions to setup your solid mechanics problem in the master... it's effectively doing this step for you (adding the variables and setting those variables as the variables to use for displacements).

Also note: that MOOSE doesn't care whether the displacement field is a nonlinear variable you solve for - or an Auxiliary variable.  This is actually really handy since it makes it easy to define functional forms of the displacements (or make them constants).

**Reply by js-jixu** [reply]
Thank you, Derke. I've sucessfully applied your suggestion to my input file and it works well.
